pkg/alerts/alertsHandler/cronJobHandler.go

```diff
@@ -116,6 +117,109 @@ func updateAlertStateAndCreateAlertHistory(alertDetails *alertutils.AlertDetails
+
+	for _, alertHistory := range alertHistoryList {
+		if !alertutils.IsAlertStatePendingOrFiring(alertHistory.AlertState) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func GetLatestAlertHistory(alertId string) (*alertutils.AlertHistoryDetails, error) {
+	alertHistoryList, err := databaseObj.GetAlertHistoryByAlertID(&alertutils.AlertHistoryQueryParams{
+		AlertId:   alertId,
+		Limit:     1,
+		SortOrder: alertutils.DESC,
+	})
+	if err != nil {
+		log.Errorf("ALERTSERVICE: GetLatestAlertHistory: Error getting AlertHistory. AlertId=%v & err=%+v.", alertId, err)
+		return nil, err
+	}
+
+	if len(alertHistoryList) == 0 {
+		return nil, nil
+	}
+
+	return alertHistoryList[0], nil
+}
+
+func handleAlertCondition(alertToEvaluate *alertutils.AlertDetails, isAlertConditionMatched bool, alertDataMessage string) error {
+	if isAlertConditionMatched {
+
+		newAlertState := alertutils.Pending
+		eventDesc := alertutils.AlertPending
+
+		if shouldUpdateAlertStateToFiring(alertToEvaluate, newAlertState) {
+			newAlertState = alertutils.Firing
+			eventDesc = alertutils.AlertFiring
+		}
+
+		err := updateAlertStateAndCreateAlertHistory(alertToEvaluate, newAlertState, eventDesc)
+		if err != nil {
+			log.Errorf("ALERTSERVICE: handleAlertCondition: Error in updateAlertStateAndCreateAlertHistory. AlertState=%v, Alert=%+v & err=%+v", newAlertState, alertToEvaluate.AlertName, err)
+		}
+
+		// If the Alert State is updated to Firing, then we should send the Alert Notification.
+		// If the previous state was Firing, then the cooldown period on the Notification Handler will decide if the notification should be sent.
+		if newAlertState == alertutils.Firing {
+			err = NotifyAlertHandlerRequest(alertToEvaluate.AlertId, alertDataMessage)
+			if err != nil {
+				return fmt.Errorf("handleAlertCondition: Could not send Alert Notification. found error = %v", err)
+			}
+		}
+	} else {
+		lastAlertHistory, err := GetLatestAlertHistory(alertToEvaluate.AlertId)
+		if err != nil {
+			log.Errorf("handleAlertCondition: Error getting latest alert history. Alert=%+v & err=%+v", alertToEvaluate.AlertName, err)
+		}
+
+		err = updateAlertStateAndCreateAlertHistory(alertToEvaluate, alertutils.Normal, alertutils.AlertNormal)
+		if err != nil {
+			log.Errorf("ALERTSERVICE: handleAlertCondition: Error in updateAlertStateAndCreateAlertHistory. AlertState=%v, Alert=%+v & err=%+v", alertutils.Normal, alertToEvaluate.AlertName, err)
+		}
+
+		// If the previous state was Firing, then we should send the Alert Notification.
+		// The cooldown period on the Notification Handler will decide if the notification should be sent. So that false positives are avoided.
+		if lastAlertHistory != nil && lastAlertHistory.AlertState == alertutils.Firing {
+			err = NotifyAlertHandlerRequest(alertToEvaluate.AlertId, "The Alert State has been updated to Normal.")
+			if err != nil {
+				return fmt.Errorf("handleAlertCondition: Could not send Alert Notification. found error = %v", err)
+			}
+		}
```

Comment:
Do you also have to tell the Notification system the alert's state? I assume we'd want the cooldown to only apply to alerts in the Firing state so we still get Normal notifications immediately, but maybe I'm wrong

I think with the current `NotifyAlertHandlerRequest()` implementation, it's possible that we'd never get a Normal message sent: suppose an alert moves to Firing, so then we get a Firing notification, and then it moves back to Normal but the notification system doesn't send a message because the cooldown is active, then eventually the cooldown ends and even if we evaluate this alert again to Normal we won't send the Normal notification because now the previous state was Normal (not Firing)

Would it make sense to always send the new state to the notification system and make the notification system responsible for deciding when to send a notification?

Reply:
I guess that makes sense, so every time when the state is in `Normal`, the alert system would call the `NotifyAlertHandlerRequest` to send the notification,  similar to `Firing` state. And then The `NotificationSystem` will decide whether to send this notification or not.

But also, the user would probably not like to receive `Normal` state alert notifications all the time. It would probably be okay, only once when the state is transitioned from `Firing` to `Normal`. 

So I believe that the Notification system should keep track of the notifications it sent or at least the last Notification Alert state it sent. 
- If the current state is in `Firing` depending on the cooldown period it will send the notification.
- If the current state is in `Normal`, then depending on the cooldown period plus the Alert state when the last Notification is sent, it should send the notification. It should only send if the Alert state of last notification that it sent was `Firing` and if it is `Normal`, it should not send.

Reply:
Yeah so do you need to also send the alert state, or will sending just the `alertID` and the message like you are doing be enough information for the notification system to figure out what to do?

Reply:
Yes, I think, I will work on that, when I am implementing the cooldown period and making fixes on Notification system.